### ISSUE TITLE
Remove examples using the extract-loader

### ISSUE THREAD

**Issue by GMartigny** [webpack-contrib/html-loader#385]
Documentation Is:

- [ ] Missing
- [ ] Needed
- [x] Confusing
- [ ] Not Sure?

### Please Explain in Detail...

The `extract-loader` don't support webpack 5, which is mandatory to use `html-loader@2`. Sadly, it seems to be [unmaintained](https://github.com/peerigon/extract-loader/issues/107) and many user are [falling into the issue](https://github.com/peerigon/extract-loader/issues/102).

### Your Proposal for Changes

The `extract-loader` shouldn't be used in the documentation. Instead, maybe advice for the use of [`mini-css-extract-plugin`](https://github.com/webpack-contrib/mini-css-extract-plugin) and [`html-webpack-plugin`](https://github.com/jantimon/html-webpack-plugin).

**Comment by alexander-akait** [comment]
Yes, but I think we can implement `extract` option for this loader

**Comment by GMartigny** [comment]
Even better !
Should I change the issue description ?

**Comment by alexander-akait** [comment]
Yes, let's change it

**Comment by shoaibsharif** [comment]
Very interested to see this feature. I was trying to extract image, style and js, but keep getting error on those.

**Comment by tremby** [comment]
Is there a way to make extraction work at present?

**Comment by ZeldOcarina** [comment]
up

**Comment by F21** [comment]
Is there currently any way or a work around to export to HTML files? The instructions in the [readme](https://github.com/webpack-contrib/html-loader#export-into-html-files) don't seem to work, probably due to `extract-loader` being unmaintained and incompatible.

My use-case is pretty simple. I have a bunch of [mjml](https://mjml.io) files and I want to use `html-loader` to parse each mjml file and and replace image source urls and srcsets to point to image assets that have been minified and renamed into a hashed filename to be deployed on a CDN. I want to use the `html-loader` to parse the files, insert the correct urls into the file and export the file back as `mjml`. I was able use `html-loader`'s custom tags configuration to deal with the parsing and rewriting of URLs, but there doesn't seem to be anyway to export the result as an html file.

**Comment by pseusys** [comment]
What is the progress?

**Comment by alexander-akait** [comment]
Somebody can provide example how you use `html-loader` and need extract, I have two solutions here, but they are different and can be used for different things, so I want to see the most popular usage

**Comment by StudioMaX** [comment]
A simple example from Webpack 4:
```js
            {
                test: /\.ejs$/i,
                exclude: /node_modules/,
                use: [
                    {
                        loader: 'ejs-loader',
                        options: {
                            variable: 'data',
                        }
                    },
                    'extract-loader',
                    'html-loader',
                ]
            }
```
- use `html-loader` to replace the paths to the assets in the EJS template
- return back to text version of EJS template
- use `ejs-loader` to convert a text template into an object (in this case, the EJS template engine from lodash is used)

`ejs-loader` is used last because exported object from the EJS templating engine is imported somewhere, and is called with data from the backend.
Something like this:
```ejs
// template.ejs
Hello, <%- data.name %>!
<% _.forEach(data.items, function (item) { %>
<div>Row - <%- item %></div>
<% }) %>
```
```js
// view.js
import helloTemplate from 'template.ejs';

let response = {name: 'John', items: ['foo', 'bar']};
$('.body').append(helloTemplate(response));
```
But most likely this is still unrealizable in the latest version of `html-loader`, since EJS template is not a valid HTML and contains specific tags.

**Comment by pseusys** [comment]
I have a `html` file in my targets and would like to process it, save to file and extract images (`img` `src` tags) to separate directory.

```html
<html>
<head>
    <link href="main.css" type="text/css" rel="stylesheet">
</head>
<body>
    <img src="hi.jpg">
</body>
</html>
```

Solution, described in `extract-loader` readme is not working (anymore):
```typescript
            },
            {
                test: /\.html$/,
                type: 'asset/resource',
                use: [
                    'extract-loader',
                    {
                        loader: "html-loader",
                        options: {
                            esModules: false
                        }
                    }
                ]
            },

```
Apparently, `extract-loader` tries to extract `hi.jpg` as a ES module and fails.

**Comment by alexander-akait** [comment]
@pseusys Why don't use `main.css` as entry point (it will be better optimized and compressed)?

@StudioMaX Weird, I think `ejs-loader` can (should) support `src`/`srcset`/etc resolving

**Comment by pseusys** [comment]
@alexander-akait because image is referenced as `src` attribute of `img` tag in html, and not in CSS.
As far as I know, one cannot set `src` via CSS. And I don't want to change that too...

**Comment by alexander-akait** [comment]
@pseusys In this case you lose assets optimizations...

**Comment by pseusys** [comment]
@alexander-akait I thought it would be easier to process the assets by extracting them from HTML files.
However you're right, if it's not possible I think I should consider replacing <img> tags with e.g. <a> tags with `background-image` property set via CSS.

**Comment by alexander-akait** [comment]
I am think thinking you `ejs-loader` should support `src`/etc resolving and create assets for them not only compile them, anyway you can try:
1. Use HtmlWebpackPlugin with the template option and set `ejs` as transformer
2. Use CopyWebpackPlugin and use the transform function too so you don't need to extract them and they will be optimized in the production build, but here limitations - no resolving, but you can pass public path there and replace, for example `<img src="{%public_path%}/images/image.png">`

Or you can even combine these approaches.

Using `import something from "./test.html"` usually means you want to use HTML template in runtime

**Comment by mbibko** [comment]
Any progress on this Feature?
In webpack 4 I use html-loader and then export HTML into their own .html file.
```
    module: {
      rules: [
        {
          test: /\.(png|svg|jpg)$/,
          use: [
            {
              loader: "file-loader",
              options: {
                context: "src",
                name: "[path][name].[ext]",
              },
            },
          ]
        },
        {
          test: /\.html$/,
          use: [
            {
              loader: "file-loader",
              options: {
                name: "[name].html",
              },
            },
            {
              loader: "extract-loader",
            },
            {
              loader: "html-loader",
              options: {
                minimize: false,
                attributes: {
                  urlFilter: (attribute, value) => !/\.(js|css)$/.test(value),
                }
              },
            },
          ],
        },
      ],
    }
```
My source HTML is:
```
<!doctype html>
<html lang="en">
<head>
  <link rel="stylesheet" href="main.css">
</head>
<body>
  <img src="./img/3.png" alt="">
  <script src="index.js"></script>
</body>
</html>
```
My entry index.js is:
```
import ("./index.html")
```

**Comment by webdiscus** [comment]
@GMartigny @mbibko @pseusys 

You can try to use the new universal [html-bundler-webpack-plugin](https://github.com/webdiscus/html-bundler-webpack-plugin).

**Note:**
- an `entry point` is an HTML template
- all source script and style files can be specified directly in HTML, JS and CSS will be extracted automatically
- all source assets defined in standard attributes (href, src, srcset, etc) will be automatically resolved and extracted
- you can use any template engine such as [Eta](https://eta.js.org/), [EJS](https://ejs.co/), [Handlebars](https://handlebarsjs.com/), [Nunjucks](https://mozilla.github.io/nunjucks/), [LiquidJS](https://github.com/harttle/liquidjs) and others without additional loaders
- defaults, supported EJS-like syntax, no additional loaders required

**Profit**
You specify all the source scripts and styles in one right place (in HTML), instead of defining them in many places: defining JS files in Webpack Entry, importing SCSS into a JS file.

### Simple usage example

For example, there is _./src/views/home/index.html_:

```html
<html>
<head>
  <!-- load source styles here -->
  <link href="./style.scss" rel="stylesheet">
  <!-- load source scripts here and/or in body -->
  <script src="./main.js" defer="defer"></script>
</head>
<body>
  <h1>Hello World!</h1>
  <img src="./logo.png">
</body>
</html>
```

The generated HTML contains the output filenames of the processed source files, while the script and link tags remain in place:
```html
<html>
<head>
  <link href="/assets/css/style.05e4dd86.css" rel="stylesheet">
  <script src="/assets/js/main.f4b855d8.js" defer="defer"></script>
</head>
<body>
  <h1>Hello World!</h1>
  <img src="/assets/img/logo.58b43bd8.png">
</body>
</html>
```

Add the HTML templates in the `entry` option (syntax is identical to [Webpack entry](https://webpack.js.org/configuration/entry-context/#entry)):

```js
const HtmlBundlerPlugin = require('html-bundler-webpack-plugin');

module.exports = {
  plugins: [
    new HtmlBundlerPlugin({
      entry: {
        // define all your templates here, the syntax is the same as Webpack entry
        index: 'src/views/index.html', // => dist/index.html
        // 'pages/about': 'src/views/about/index.html', // => dist/pages/about.html
        // ...
      },
      js: {
        // output filename of extracted JS from source script loaded in HTML via `<script>` tag
        filename: 'assets/js/[name].[contenthash:8].js',
      },
      css: {
        // output filename of extracted CSS from source style loaded in HTML via `<link>` tag
        filename: 'assets/css/[name].[contenthash:8].css',
      },
    }),
  ],

  module: {
    rules: [
      // styles
      {
        test: /\.(css|sass|scss)$/,
        use: ['css-loader', 'sass-loader'],
      },
      // images
      {
        test: /\.(ico|png|jp?g|svg)/,
        type: 'asset/resource',
        generator: {
          filename: 'assets/img/[name].[hash:8][ext][query]',
        },
      },
    ],
  },
};
```

**Comment by tremby** [comment]
That all sounds fantastic.
